bin/olly.ml

```diff
@@ -6,14 +6,23 @@ open Cmdliner
    .fxt). Fuchsia is a binary format viewable in Perfetto. *)
 type trace_format = Json | Fuchsia
 
+type ts = { mutable start_time : float; mutable end_time : float }
+
 let total_gc_time = ref 0
-let start_time = ref 0.0
-let end_time = ref 0.0
+let wall_time = { start_time = 0.; end_time = 0. }
+let domain_cpu_times = Hashtbl.create 128
+let domain_gc_times = Array.make 128 0
 
-let lifecycle _domain_id _ts lifecycle_event _data =
+let lifecycle domain_id _ts lifecycle_event _data =
   match lifecycle_event with
-  | Runtime_events.EV_RING_START -> start_time := Unix.gettimeofday ()
-  | Runtime_events.EV_RING_STOP -> end_time := Unix.gettimeofday ()
+  | Runtime_events.EV_RING_START -> wall_time.start_time <- Unix.gettimeofday ()
```

Comment:
Is the idea that for domain 0 we measure the time with RING_START and RING_STOP, and DOMAIN_SPAWN and DOMAIN_TERMINATE for other domains?

Reply:
That's the idea. Currently, `EV_DOMAIN_SPAWN` and `EV_DOMAIN_TERMINATE` are recorded at `caml_domain_spawn` and `domain_terminate` respectively, so they're not logged for Domain 0.

Reply:
Ok. 

Have you run this PR on the solver-service PR and does the observed numbers match the results observed using `perf`. Essentially, can we independently check that the results produced by the PR is correct by checking the results against what's observed in `perf`.

Reply:
Any updates on this? Just want to make sure that our reported results can be validated independently.

Reply:
Yes, we should validate the results against perf/`usr/bin/time`. I'll post the numbers here.

I got sucked into a heisenbug I found on this PR. I'll put it somewhere when I get it to segfault reliably. My hunch is that there's some bad interaction happening with lifecycle events + core, I'm not able to pin point the exact reason yet. I'm almost certain the bug is not in olly, it happens even when we do very little stuff at domain_spawn and domain_terminate. I've looked at the runtime code and don't see anything odd there either. Will dig into this.

Reply:
I've put the bug reproduction here - https://github.com/Sudha247/olly-bug-repro. Will update the PR later this week to avoid the bug.

cc @sadiqj, in case you see something obviously off.